### ISSUE TITLE
Category rule - Approver field is not empty after the approver is removed from the workspace

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#49574]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
 
**Version Number:** 9.0.39-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** applausetester+kh010901@applause.expensifail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
Precondition:
- Rules are enabled.
- "Add approvals" is enabled in Workflows.
- Workspace has at least one member other than yourself.

1. Go to staging.new.expensify.com
2. Go to workspace settings > Categories..
3. Click on any category.
4. Click Approver.
5. Select a member as the approver.
6. Go to Members.
7. Remove the member from the workspace.
8. Go back to the category rule of the category in Step 3.

## Expected Result:
Approver field will be empty because the approver is no longer a workspace member.

## Actual Result:
Approver field still shows the removed member when the approver is no longer a workspace member.

## Workaround:
Unknown

## Platforms:

- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/d169ef74-d918-435d-832f-8fe111f02d22

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by IuliiaHerets** [comment]
We think that this bug might be related to #wave-control

**Comment by IuliiaHerets** [comment]
@mallenexpensify  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by mallenexpensify** [comment]
Checking with Tom about quick

**Comment by mallenexpensify** [comment]
Posted and bumped in #wave-control room https://expensify.slack.com/archives/C06ML6X0W9L/p1727482508265999?thread_ts=1727123059.243429&cid=C06ML6X0W9L

**Comment by mallenexpensify** [comment]
[Per Dylan](https://expensify.slack.com/archives/C06ML6X0W9L/p1727685299292309?thread_ts=1727123059.243429&cid=C06ML6X0W9L), Put on hold pending the beta being built for it
- https://github.com/Expensify/Expensify/issues/431466

**Comment by mallenexpensify** [comment]
On hold

**Comment by mallenexpensify** [comment]
Was able to reproduce 
<img width="778" alt="SnagitHelper2024 2024-11-06 13 45 44" src="https://github.com/user-attachments/assets/7e662444-16d0-4ceb-ab59-b2e635d4284f">


and

<img width="402" alt="SnagitHelper2024 2024-11-06 13 46 16" src="https://github.com/user-attachments/assets/6ce49dd3-db69-44fd-98ca-ed63e88af3f3">


@mollfpr, can you review the proposal above? Thx

**Comment by mallenexpensify** [comment]
@zanyrenney I'm off the next week, can you please keep an eye on this issue til I'm back? Thx

**Comment by zanyrenney** [comment]
Sure thing @mallenexpensify !

**Comment by zanyrenney** [comment]
Seeing as @mallenexpensify successfully reproduced this, I am removing the label.

**Comment by zanyrenney** [comment]
bumping @mollfpr to review the proposal above.

**Comment by mollfpr** [comment]
>We don't reset the approval rule to be empty after deleting a member.

@abzokhattab I agree with RCA, but why is the solution not related to resetting the rules?

**Comment by mallenexpensify** [comment]
@mollfpr , 👀 above plz.  Thanks for watching over Zany, I'm back!

**Comment by mollfpr** [comment]
>We don't reset the approval rule to be empty after deleting a member.

@abzokhattab, I'm still not onboard with the solution. The root cause is in our BE response not updating the policy `rules.approvalRules` after removing the member. So I think the proper solution is:
1. BE: Fixing the API `DeleteMembersFromWorkspace` response to have the `rules.approvalRules` updated with the employee list
2. FE: Follow up with the same approach for the OFFLINE

That's it, no need to update `getCategoryApproverRule` since we got the correct data for `rules.approvalRules`.

**Comment by mollfpr** [comment]
Ohyeah, you already have that in the alternative solution my bad 🙏 

The alternative solution looks good to me and it's pretty much the same as the [strategy](https://github.com/Expensify/App/issues/49574#issuecomment-2484659473) I suggest. I think we can move forward from here and discuss the BE changes with the selected engineer.

🎀 👀 🎀 C+ reviewed!

**Comment by francoisl** [comment]
Yeah let's go with the backend + front-end solution. The backend should definitely clean up the `approvalRules` array in this case.

I'm done with the bulk of the backend changes but still need to add some tests. Heads up that I'm going to be OOO for the rest of the week, but feel free to start working on the front-end changes for the optimistic and failure data in the meantime.

**Comment by francoisl** [comment]
Will do. The backend change PR is in review now.

**Comment by francoisl** [comment]
@abzokhattab the backend changes are now live.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by mallenexpensify** [comment]
Looks like the PR is moving along
- https://github.com/Expensify/App/pull/53383#pullrequestreview-2507550198

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by mallenexpensify** [comment]
PR is on staging
https://github.com/Expensify/App/pull/53383#issuecomment-2551348519

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Third week)

**Comment by mallenexpensify** [comment]
Contributor:  @abzokhattab paid $250 via Upwork 
Contributor+: @mollfpr due $250 via NewDot  

@mollfpr it doesn't look like the BZ checklist was auto-added.  Can you fill it out please?  

# BugZero Checklist:

- [ ] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [ ] 1z. Other:

Where bug was reported:
  - [ ] 2a. Reported on production (eg. bug slipped through the normal regression and PR testing process on staging)
  - [ ] 2b. Reported on staging (eg. found during regression or PR testing)
  - [ ] 2d. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [ ] 3d. QA
  - [ ] 3z. Other:

</details>

- [ ] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment:

- [ ] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion:

- [ ] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

<details>
<summary>Regression Test Proposal Template</summary>


- [ ] **[BugZero Assignee]** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


-

### Test:


1.

Do we agree 👍 or 👎


</details>

**Comment by mallenexpensify** [comment]
@mollfpr 👀 plz above.  Thx

**Comment by mollfpr** [comment]
# BugZero Checklist:

- [x] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [x] 1d. Missing an edge case on Backend and Frontend
  - [ ] 1z. Other:

Where bug was reported:
  - [x] 2a. Reported on production (eg. bug slipped through the normal regression and PR testing process on staging)
  - [ ] 2b. Reported on staging (eg. found during regression or PR testing)
  - [ ] 2d. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [x] 3d. QA
  - [ ] 3z. Other:

</details>

- [x] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment: No offending PR. We've missed a case on called `DeleteMembersFromWorkspace` where it's should update the approval rules too.

- [x] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion: Not a critical issue.

- [x] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.


- [ ] **[BugZero Assignee]** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


- Rules are enabled.
- "Add approvals" is enabled in Workflows.
- Workspace has at least one member other than yourself.


### Test:


1. Go to workspace settings > Categories..
2. Click on any category.
3. Click Approver.
4. Select a member as the approver.
5. Go to Members.
6. Find the member selected as approver before and remove the member from the workspace.
7. Go back to the category rule of the category in Step 3.
8. Ensure that the approval item is cleared

Do we agree 👍 or 👎

**Comment by mallenexpensify** [comment]
Thanks @mollfpr , test case created, should be all good here! 
- https://github.com/Expensify/Expensify/issues/457916

**Comment by garrettmknight** [comment]
$250 approved for @mollfpr
